src/Providers/Maps/Google/Features/Shape.ts

```diff
@@ -0,0 +1,144 @@
+            const distanceBetweenPoints =
+                google.maps.geometry.spherical.computeDistanceBetween(
+                    this._markerCoordinates,
+                    circleCenter
+                );
+
+            this._isInsideShape = distanceBetweenPoints <= circleRadius;
+            this._returnObjSuccess = true;
+            this._returnObjCode = OSFramework.Maps.Enum.Success.code;
+            this._returnObjMessage = this._isInsideShape.toString();
+        }
+
+        // Method to apply the default calculations for shapes
+        private _shapeDefault(): void {
+            this._isInsideShape = google.maps.geometry.poly.containsLocation(
+                this._markerCoordinates,
+                this._map.getShape(this._shape.widgetId).provider
+            );
+            this._returnObjSuccess = true;
+            this._returnObjCode = OSFramework.Maps.Enum.Success.code;
+            this._returnObjMessage = this._isInsideShape.toString();
+        }
+
+        // Method for shape rectangle exception
+        private _shapeRectangle(): void {
+            this._isInsideShape = this._map
+                .getShape(this._shape.widgetId)
+                .provider.getBounds()
+                .contains(this._markerCoordinates);
+            this._returnObjSuccess = true;
+            this._returnObjCode = OSFramework.Maps.Enum.Success.code;
+            this._returnObjMessage = this._isInsideShape.toString();
+        }
+
+        public build(): void {
+            //
+        }
+
+        public containsLocation(
+            mapId: string,
+            shapeId: string,
+            pointCoordinates: string,
+            coordinatesList: string
+        ): OSFramework.Maps.OSStructures.ReturnMessage {
+            this._map = OutSystems.Maps.MapAPI.MapManager.GetMapById(mapId);
+            // Set the default return message to prevent different else if's
+            this._returnObjSuccess = false;
+            this._returnObjCode =
+                OSFramework.Maps.Enum.ErrorCodes.CFG_InvalidMapId;
+
+            // Check if map exists
+            if (this._map) {
+                this._isInsideShape = false;
+
+                // Convert to JSON the marker location before to convert to google format
+                this._markerCoordinates = JSON.parse(pointCoordinates);
+                this._markerCoordinates = new google.maps.LatLng(
+                    this._markerCoordinates.Lat,
+                    this._markerCoordinates.Lng
+                );
+
+                // Check if marker is inside shape based on shape element
+                if (shapeId) {
+                    this._shape =
+                        OutSystems.Maps.MapAPI.ShapeManager.GetShapeById(
+                            shapeId
+                        );
+
+                    // Check if shape contains marker based on shape type
+                    switch (this._shape.type) {
```

Comment:
what happens if the this._shape is undefined or null?

Reply:
They will go to the default